### PR TITLE
Changes to support RDMA Tests for VOQ Chassis

### PR DESCRIPTION
**Pull request by YatishSVC**
### Description of PR

Summary:
Fixes # (issue)
17581 https://github.com/sonic-net/sonic-mgmt/issues/17581
### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ x] Test case improvement

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411

#### What is the motivation for this PR?
In case of a VOQ Chassis, the loss happens on the ingress side instead of the egress side which happens in case of Non- VOQ Chassis.
#### How did you do it?
By adding a condition for VOQ Chassis.
#### How did you verify/test it?
Ran it in Microsoft Lab.

### CODE REVIEW

**Comment by YatishSVC** [comment]
@amitpawar12 @sdszhang @vmittal-msft For vis

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 1]
```diff
@@ -134,15 +134,27 @@ def run_m2o_fluctuating_lossless_test(api,
...
+        pkt_drop_1_ingress = get_interface_stats(
+            ingress_dut1, ingress_port1
+        )[ingress_dut1.hostname][ingress_port1]['rx_drp']
+        pkt_drop_2_ingress = get_interface_stats(
+            ingress_dut2, ingress_port2
+        )[ingress_dut2.hostname][ingress_port2]['rx_drp']
+        total_pkt_drop_ingress = pkt_drop_1_ingress + pkt_drop_2_ingress
+        drop_percentage = (100 * total_pkt_drop_ingress) / total_rx_pkts
+
+        pytest_assert(abs(drop_percentage - 8) < 4, "FAIL: Drop packets must be around 8 percent")
```
Comment (arlakshm) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:151]: what does this `4` mean?

Comment (YatishSVC) [reply tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:151]: > what does this `4` mean?

This is for margin of +-4 around 8%

**Comment by sdszhang** [comment]
@amitpawar12 for review.

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 2]
```diff
@@ -134,15 +134,25 @@ def run_m2o_fluctuating_lossless_test(api,
...
+            ingress_dut2, ingress_port2
+        )[ingress_dut2.hostname][ingress_port2]['rx_drp']
+        total_pkt_drop_ingress = pkt_drop_1_ingress + pkt_drop_2_ingress
+        drop_percentage = (100 * total_pkt_drop_ingress) / total_rx_pkts
+
+    else:
+        pkt_drop = get_interface_stats(egress_duthost, dut_tx_port)[egress_duthost.hostname][dut_tx_port]['tx_drp']
+        drop_percentage = (100 * pkt_drop) / total_rx_pkts
+
+    pytest_assert(abs(drop_percentage - 8) < 4, "FAIL: Drop packets must be around 8 percent")
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:155]: Will this verification logic apply to non voq chassis as well ?

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py [hunk 3]
```diff
@@ -414,4 +424,7 @@ def verify_m2o_fluctuating_lossless_result(rows,
             pytest_assert(int(row.loss) == 0, "FAIL: {} must have 0% loss".format(row.name))
         elif 'Background Flow' in row.name:
             background_loss += float(row.loss)
-    pytest_assert(round(background_loss/4) == 10, "Each Background Flow must have an avg of 10% loss ")
+    pytest_assert(
+        8 <= round(background_loss / 4) <= 12,
+        "Each Background Flow must have an avg of 10% loss (Allowed ±2%)"
+    )
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py:430]: Same comment as above.

File: tests/snappi_tests/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py [hunk 4]
```diff
@@ -143,15 +143,25 @@ def run_pfc_m2o_oversubscribe_lossless_lossy_test(api,
...
+            ingress_dut2, ingress_port2
+        )[ingress_dut2.hostname][ingress_port2]['rx_drp']
+        total_pkt_drop_ingress = pkt_drop_1_ingress + pkt_drop_2_ingress
+        drop_percentage = (100 * total_pkt_drop_ingress) / total_rx_pkts
+
+    else:
+        pkt_drop = get_interface_stats(egress_duthost, dut_tx_port)[egress_duthost.hostname][dut_tx_port]['tx_drp']
+        drop_percentage = (100 * pkt_drop) / total_rx_pkts
+
+    pytest_assert(abs(drop_percentage - 5) < 4, 'FAIL: Drop packets must be around 5 percent')
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_oversubscribe_lossless_lossy_helper.py:164]: Same comment as above.

File: tests/snappi_tests/pfc/files/m2o_oversubscribe_lossy_helper.py [hunk 5]
```diff
@@ -144,14 +144,25 @@ def run_pfc_m2o_oversubscribe_lossy_test(api,
...
+            ingress_dut2, ingress_port2
+        )[ingress_dut2.hostname][ingress_port2]['rx_drp']
+        total_pkt_drop_ingress = pkt_drop_1_ingress + pkt_drop_2_ingress
+        drop_percentage = (100 * total_pkt_drop_ingress) / total_rx_pkts
+
+    else:
+        pkt_drop = get_interface_stats(egress_duthost, dut_tx_port)[egress_duthost.hostname][dut_tx_port]['tx_drp']
+        drop_percentage = (100 * pkt_drop) / total_rx_pkts
+
+    pytest_assert(abs(drop_percentage - 10) < 2, 'FAIL: Drop packets must be around 10 percent')
```
Comment (vmittal-msft) [inline tests/snappi_tests/pfc/files/m2o_oversubscribe_lossy_helper.py:165]: same comment as above

File: tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py (hunk 2 above)
Comment (YatishSVC) [reply tests/snappi_tests/pfc/files/m2o_fluctuating_lossless_helper.py]: Hi @vmittal-msft,
 Only the pytest_assert was kept common between the voq and non-voq platforms, but I am updating them to keep them separate for voq and non-voq platforms.

**Comment by amitpawar12** [comment]
@yatishkoul , @YatishSVC-

There are two aspects to this PR:
1. Change the drop_percentage to be calculated for VOQ chassis by capturing pkt_drops on ingress interfaces.
2. Modification of pass-fail criteria for the drop_percentage. 

IMHO, these should be treated as two independent PRs. If required, we can always determine the order in which PRs will be merged in, which I don't think is the issue in this case.

For second item, I think there should be broader agreement between concerned teams about what should be range of the drop packets, ensuring that there is no false-pass introduced. I will recommend addressing ONLY item#1 in this PR and raise a new PR for item#2.

Tagging @vmittal-msft , @rawal01 to see if they think otherwise.

Hope you will agree to this.

Thanks,
-A

**Comment by YatishSVC** [comment]
> @yatishkoul , @YatishSVC-
> 
> There are two aspects to this PR:
> 
> 1. Change the drop_percentage to be calculated for VOQ chassis by capturing pkt_drops on ingress interfaces.
> 2. Modification of pass-fail criteria for the drop_percentage.
> 
> IMHO, these should be treated as two independent PRs. If required, we can always determine the order in which PRs will be merged in, which I don't think is the issue in this case.
> 
> For second item, I think there should be broader agreement between concerned teams about what should be range of the drop packets, ensuring that there is no false-pass introduced. I will recommend addressing ONLY item#1 in this PR and raise a new PR for item#2.
> 
> Tagging @vmittal-msft , @rawal01 to see if they think otherwise.
> 
> Hope you will agree to this.
> 
> Thanks, -A

Hi @amitpawar12 , 
Yes, that would be the right way of doing this, so, currently the pr has only changes for the first item that is VOQ and Non Voq Changes.
Can you please review these changes and let me know if anything else is required here?

**Review by vmittal-msft — APPROVED** [review]

**Comment by vmittal-msft** [comment]
> > @yatishkoul , @YatishSVC-
> > There are two aspects to this PR:
> > 
> > 1. Change the drop_percentage to be calculated for VOQ chassis by capturing pkt_drops on ingress interfaces.
> > 2. Modification of pass-fail criteria for the drop_percentage.
> > 
> > IMHO, these should be treated as two independent PRs. If required, we can always determine the order in which PRs will be merged in, which I don't think is the issue in this case.
> > For second item, I think there should be broader agreement between concerned teams about what should be range of the drop packets, ensuring that there is no false-pass introduced. I will recommend addressing ONLY item#1 in this PR and raise a new PR for item#2.
> > Tagging @vmittal-msft , @rawal01 to see if they think otherwise.
> > Hope you will agree to this.
> > Thanks, -A
> 
> Hi @amitpawar12 , Yes, that would be the right way of doing this, so, currently the pr has only changes for the first item that is VOQ and Non Voq Changes. Can you please review these changes and let me know if anything else is required here?

I agree. btw having drop %age rate of 4-12% is too wide and need to be discussed. Approving as changes are made based on comment.

**Review by arlakshm — APPROVED** [review]

**Review by arlakshm — APPROVED** [review]
